### PR TITLE
Collected small changes and fixes

### LINKED ISSUE #3799: [BUG] Misalignment of Thermo Data in dummped YAML file
**Summary**

When comparing the thermo output from the LAMMPS log file to the thermo output in the YAML dump file, I've noticed a difference for the initial timestep (0). In the YAML file, the thermo data for timestep 0 appears to be invalid or meaningless, and does not align with the thermo data from the LAMMPS log file.

As can be observed from the following snipptes, the thermo data for timestep 1 in the YAML file matches the data for timestep 0 in the LAMMPS log file. This pattern continues, causing all thermo data in the YAML file to appear one timestep behind the data in the LAMMPS log file. This discrepancy is causing significant confusion when analyzing the results.

Here's the snippet of the thermo data from the LAMMPS log file for the first few timesteps:
```
keywords: ['Step', 'PotEng', 'KinEng', 'TotEng', 'Temp', 'Press', 'Volume', 'Density', 'c_press[1]', 'c_press[2]', 'c_press[3]', 'c_press[4]', 'c_press[5]', 'c_press[6]', ]
data:
  - [0, -479421.691781166, 0, -479421.691781166, 0, 7864.6559730265, 512, 0.5842697004374, 7720.76660099825, 11152.0949780855, 4721.10633999577, 1861.50214355324, 1598.1849785235, -415.12650213297, ]
  - [1, -479421.702029192, 0.0102364454473237, -479421.691792747, 0.118417769523395, 7835.45786565537, 512, 0.5842697004374, 7691.68586639822, 11120.8612768112, 4693.82645375669, 1858.27672454755, 1594.21198346326, -410.512686839785, ]
  - [2, -479421.7325883, 0.0407611223356872, -479421.691827178, 0.471534891198419, 7748.08050985201, 512, 0.5842697004374, 7604.65988968182, 11027.4111059135, 4612.17053396075, 1848.63117983296, 1582.32645091386, -396.700177502109, ]
  - [3, -479421.782908429, 0.0910248771390507, -479421.691883551, 1.05299861923903, 7603.17544035678, 512, 0.5842697004374, 7460.34501669082, 10872.4869496582, 4476.69435472129, 1832.66461502789, 1562.61735161485, -373.785220639499, ]
  - [4, -479421.852088677, 0.160128297181147, -479421.691960379, 1.85240432212028, 7401.79815680349, 512, 0.5842697004374, 7259.81679407431, 10657.2768161709, 4288.30086016521, 1810.50727012551, 1535.25774010898, -341.898159687684, ]
...
```

And here's the corresponding output in the YAML dump file for the same timesteps:
```
creator: LAMMPS
timestep: 0
units: real
time: 0
natoms: 30
boundary: [ p, p, p, p, p, p, ]
thermo:
  - keywords: [ Step, PotEng, KinEng, TotEng, Temp, Press, Volume, Density, Pxx, Pyy, Pzz, Pxy, Pxz, Pyz, ]
  - data: [ 0, -nan, 0, -nan, 1.9090011553e-313, 5.43223878283e-312, 512, 0.5842697004374003, 4.6600098255045e-310, 0, 4.6601149565388e-310, 4.787190529701563e+180, 1.8155354104984654e-134, 5.817149395556409e+180, ]
...
---
creator: LAMMPS
timestep: 1
units: real
time: 0.1
natoms: 30
boundary: [ p, p, p, p, p, p, ]
thermo:
  - keywords: [ Step, PotEng, KinEng, TotEng, Temp, Press, Volume, Density, Pxx, Pyy, Pzz, Pxy, Pxz, Pyz, ]
  - data: [ 1, -479421.6917811662, 0, -479421.6917811662, 0, 7864.655973026495, 512, 0.5842697004374003, 7720.766600998249, 11152.094978085463, 4721.106339995775, 1861.5021435532385, 1598.1849785234965, -415.12650213297024, ]
...
---
creator: LAMMPS
timestep: 2
units: real
time: 0.2
natoms: 30
boundary: [ p, p, p, p, p, p, ]
thermo:
  - keywords: [ Step, PotEng, KinEng, TotEng, Temp, Press, Volume, Density, Pxx, Pyy, Pzz, Pxy, Pxz, Pyz, ]
  - data: [ 2, -479421.7020291921, 0.0102364454473237, -479421.69179274666, 0.11841776952339539, 7835.457865655372, 512, 0.5842697004374003, 7691.6858663982175, 11120.861276811209, 4693.826453756694, 1858.276724547549, 1594.2119834632642, -410.5126868397848, ]
...
---
creator: LAMMPS
timestep: 3
units: real
time: 0.3
natoms: 30
boundary: [ p, p, p, p, p, p, ]
thermo:
  - keywords: [ Step, PotEng, KinEng, TotEng, Temp, Press, Volume, Density, Pxx, Pyy, Pzz, Pxy, Pxz, Pyz, ]
  - data: [ 3, -479421.7325883004, 0.04076112233568721, -479421.69182717806, 0.471534891198419, 7748.0805098520095, 512, 0.5842697004374003, 7604.659889681816, 11027.41110591347, 4612.170533960747, 1848.6311798329627, 1582.3264509138571, -396.7001775021091, ]
...
---
creator: LAMMPS
timestep: 4
units: real
time: 0.4
natoms: 30
boundary: [ p, p, p, p, p, p, ]
thermo:
  - keywords: [ Step, PotEng, KinEng, TotEng, Temp, Press, Volume, Density, Pxx, Pyy, Pzz, Pxy, Pxz, Pyz, ]
  - data: [ 4, -479421.78290842863, 0.09102487713905066, -479421.6918835515, 1.0529986192390275, 7603.17544035678, 512, 0.5842697004374003, 7460.345016690816, 10872.486949658234, 4476.6943547212895, 1832.6646150278937, 1562.6173516148476, -373.7852206394987, ]
...
```

**LAMMPS Version and Platform**

Large-scale Atomic/Molecular Massively Parallel Simulator - 28 Mar 2023 - Development

**To reproduce**

```
compute        press all pressure thermo_temp

dump           myDump all yaml 1 dump.yaml id type x y z fx fy fz  # c_atomicenergies
dump_modify    myDump time yes units yes thermo yes format float "% 10.10e"

thermo_style   custom step pe ke etotal temp press vol density c_press[*]
thermo_modify line yaml format none
thermo         1
```

### PR DESCRIPTION
**Pull request by akohlmey**
**Summary**

This pull request combines multiple small changes and bugfixes.

**Related Issue(s)**

Addresses the issue reported here: https://matsci.org/t/python-gather-bonds-error/48618/
Fixes #3799
Addresses performance regression reported here: https://matsci.org/t/performance-degradation-in-neighbor-list-construction-in-openmp-package/48973

**Author(s)**

Axel Kohlmeyer, Temple U

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No known issues.

**Implementation Notes**

The following individual changes are included:
- make sure the buffer for gathering bond/angle/dihedral/improper information is non-null and thus avoid a segfault
- correct minor inconsistencies in water model howto examples
- improve some error messages
- swap order of creating thermo and dump file output in order for dump styles that include thermo data to be in sync
- refactor how dump styles can access thermo data by creating an explicit cache of computed values
- add library interface API to access the cached thermo data
- fix bug when Kokkos border comm is on host
- fix performance regression when using multi-threaded neighbor list build in OPENMP package
- add support for atom type wildcards to fix reaxff/species
- use advanced fmtlib formatting to create column aligned output for fix reaxff/species instead of depending on tabs

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.

### CODE REVIEW

**Comment by akohlmey** [comment]
Note: the changes to the Output class may need extra testing. It has passed all tests I can trigger from GitHub.

**Review by athomps — APPROVED** [review]
I approve.

**Review by stanmoore1 — APPROVED** [review]

File: src/NETCDF/dump_netcdf_mpiio.cpp [hunk 1]
```diff
@@ -594,20 +593,17 @@ void DumpNetCDFMPIIO::write()
   NCERR( ncmpi_begin_indep_data(ncid) );
 
   if (thermo) {
-    Thermo *th = output->thermo;
-    for (int i = 0; i < th->nfield; i++) {
-      th->call_vfunc(i);
+    auto keywords = output->thermo->get_keywords();
+    auto fields = output->thermo->get_fields();
```
Comment (rbberger) [inline src/NETCDF/dump_netcdf_mpiio.cpp]: These should both be `const auto &` to avoid copies.

File: src/NETCDF/dump_netcdf_mpiio.cpp [hunk 2]
```diff
@@ -422,18 +423,16 @@ void DumpNetCDFMPIIO::openfile()
...
-        } else if (th->vtype[i] == Thermo::INT) {
-          NCERRX( ncmpi_def_var(ncid, th->keyword[i].c_str(), NC_INT, 1, dims, &thermovar[i]), th->keyword[i].c_str() );
-        } else if (th->vtype[i] == Thermo::BIGINT) {
-#if defined(LAMMPS_SMALLBIG) || defined(LAMMPS_BIGBIG)
-          NCERRX( ncmpi_def_var(ncid, th->keyword[i].c_str(), NC_INT64, 1, dims, &thermovar[i]), th->keyword[i].c_str() );
-#else
-          NCERRX( ncmpi_def_var(ncid, th->keyword[i].c_str(), NC_LONG, 1, dims, &thermovar[i]), th->keyword[i].c_str() );
-#endif
+      auto fields = output->thermo->get_fields();
+      auto keywords = output->thermo->get_keywords();
```
Comment (rbberger) [inline src/NETCDF/dump_netcdf_mpiio.cpp]: Same here.

File: src/NETCDF/dump_netcdf_mpiio.cpp [hunk 3]
```diff
@@ -318,9 +318,10 @@ void DumpNetCDFMPIIO::openfile()
 
     // perframe variables
     if (thermo) {
-      Thermo *th = output->thermo;
-      for (int i = 0; i < th->nfield; i++) {
-        NCERRX( ncmpi_inq_varid(ncid, th->keyword[i].c_str(), &thermovar[i]), th->keyword[i].c_str() );
+      auto keywords = output->thermo->get_keywords();
```
Comment (rbberger) [inline src/NETCDF/dump_netcdf_mpiio.cpp]: and here

File: src/NETCDF/dump_netcdf.cpp [hunk 4]
```diff
@@ -605,20 +600,17 @@ void DumpNetCDF::write()
   start[1] = 0;
 
   if (thermo) {
-    Thermo *th = output->thermo;
-    for (int i = 0; i < th->nfield; i++) {
-      th->call_vfunc(i);
+    auto keywords = output->thermo->get_keywords();
+    auto fields = output->thermo->get_fields();
```
Comment (rbberger) [inline src/NETCDF/dump_netcdf.cpp]: here

File: src/NETCDF/dump_netcdf.cpp [hunk 5]
```diff
@@ -432,22 +433,16 @@ void DumpNetCDF::openfile()
...
-          } else if (th->vtype[i] == Thermo::BIGINT) {
-#if defined(LAMMPS_SMALLBIG) || defined(LAMMPS_BIGBIG)
-            NCERRX( nc_def_var(ncid, th->keyword[i].c_str(), NC_INT64, 1, dims,
-                               &thermovar[i]), th->keyword[i].c_str() );
-#else
-            NCERRX( nc_def_var(ncid, th->keyword[i].c_str(), NC_LONG, 1, dims,
-                               &thermovar[i]), th->keyword[i].c_str() );
-#endif
+        auto fields = output->thermo->get_fields();
+        auto keywords = output->thermo->get_keywords();
```
Comment (rbberger) [inline src/NETCDF/dump_netcdf.cpp]: here

File: src/NETCDF/dump_netcdf.cpp [hunk 6]
```diff
@@ -320,9 +320,10 @@ void DumpNetCDF::openfile()
 
       // perframe variables
       if (thermo) {
-        Thermo *th = output->thermo;
-        for (int i = 0; i < th->nfield; i++) {
-          NCERRX( nc_inq_varid(ncid, th->keyword[i].c_str(), &thermovar[i]), th->keyword[i].c_str() );
+        auto keywords = output->thermo->get_keywords();
```
Comment (rbberger) [inline src/NETCDF/dump_netcdf.cpp]: here

File: src/EXTRA-DUMP/dump_yaml.cpp [hunk 7]
```diff
@@ -60,21 +60,26 @@ void DumpYAML::write_header(bigint ndump)
...
-      else if (th->vtype[i] == Thermo::BIGINT)
-        thermo_data += fmt::format("{}, ", th->bivalue);
+    // output thermo data only on timesteps where it was computed
+    if (update->ntimestep == th->get_timestep()) {
+
+      thermo_data += "thermo:\n  - keywords: [ ";
+      for (auto key : th->get_keywords()) thermo_data += fmt::format("{}, ", key);
+      thermo_data += "]\n  - data: [ ";
+
+      for (auto val : th->get_fields()) {
```
Comment (rbberger) [inline src/EXTRA-DUMP/dump_yaml.cpp]:
```suggestion
      for (const auto & val : th->get_fields()) {
```

File: src/EXTRA-DUMP/dump_yaml.cpp [hunk 8]
```diff
@@ -60,21 +60,26 @@ void DumpYAML::write_header(bigint ndump)
...
-        thermo_data += fmt::format("{}, ", th->dvalue);
-      else if (th->vtype[i] == Thermo::INT)
-        thermo_data += fmt::format("{}, ", th->ivalue);
-      else if (th->vtype[i] == Thermo::BIGINT)
-        thermo_data += fmt::format("{}, ", th->bivalue);
+    // output thermo data only on timesteps where it was computed
+    if (update->ntimestep == th->get_timestep()) {
+
+      thermo_data += "thermo:\n  - keywords: [ ";
+      for (auto key : th->get_keywords()) thermo_data += fmt::format("{}, ", key);
```
Comment (rbberger) [inline src/EXTRA-DUMP/dump_yaml.cpp]:
```suggestion
      for (const auto & key : th->get_keywords()) thermo_data += fmt::format("{}, ", key);
```

**Review by rbberger — CHANGES_REQUESTED** [review]
Nice addition, just avoid the copies.

File: src/NETCDF/dump_netcdf_mpiio.cpp (hunk 1 above)
Comment (akohlmey) [reply src/NETCDF/dump_netcdf_mpiio.cpp]: Shouldn't this be automatic when Thermo::get_keywords() returns a const &?

**Review by sjplimp — APPROVED** [review]

**Comment by sjplimp** [comment]
@akohlmey New thermo field access logic looks good - thanks for iterating with me on this change

File: src/NETCDF/dump_netcdf_mpiio.cpp (hunk 1 above)
Comment (rbberger) [reply src/NETCDF/dump_netcdf_mpiio.cpp]: No. That's not how `auto` works. https://godbolt.org/z/e5zPYjbdT

**Comment by akohlmey** [comment]
> New thermo field access logic looks good

Not quite done with that. Needed to make some tweaks so I can expose it to the library interface.
This is mostly done, but now the challenge it to get it into Python (and the other languages). To make it work with PyLammps instead of capturing and parsing the screen output, I may have to add an option to run a callback function whenever thermo output is created. But one step at a time...

**Review by rbberger — APPROVED** [review]

**Comment by rbberger** [comment]
@akohlmey let me know if you want me to update the Python parts.

**Comment by akohlmey** [comment]
> @akohlmey let me know if you want me to update the Python parts.

Let me see how far I get tomorrow. Next step is writing tests (and sorting out the failing ones).
Will have to think about what a good interface for python is. Perhaps returning a dict?
Then I need to propagate the function through the other language interfaces (fortran, swig, dlopen). They are all "complete" at the moment and I would like to keep it that way, especially since we are close to a stable release.

Writing tests is a good way to check whether the added interface is usable or needs more tweaks (or adding a companion function).

**Review by jtclemm — APPROVED** [review]
Looks good.

**Comment by akohlmey** [comment]
@rbberger test and lammps module wrapper are added. PyLammps is going to be more work, especially if you want to use this internally, so I didn't touch that. Had to make a few small tweaks to make the interface work well enough with just one function call added.

Now looking into swig, fortran, and dlopen versions...

**Comment by rbberger** [comment]
@akohlmey looks ok. Although I personally would have stuck with the void* cast and not use the static to keep the interface "thread-safe". but I'm not even sure that's something we're striving for anyway. `typ = cast(ptr, c_int).value` should have been enough instead of resetting the function return type. Either way, I'm ok with it.

**Comment by akohlmey** [comment]
> Although I personally would have stuck with the void* cast and not use the static to keep the interface "thread-safe".

I have tried hard and it works with ctypes, but GCC was throwing a fit when I tried to cast the void* back to an int when I was writing the unit test. The alternative would have been to use another union similar to ubuf and that would have failed on big endian machines (which are almost extinct, though) or required a maze of ifdefs to detect those. I decided that having to use a static variable in the library interface is the - by far - lesser evil.

**Comment by rbberger** [comment]
@akohlmey Probably it's just telling us that the cleaner way would be to add a second function.

**Pull request merged**